featurebyte/query_graph/transform/decompose_point.py

```diff
@@ -516,33 +548,47 @@ def _post_compute(
     ) -> DecomposePointState:
         input_node_names = self.graph.get_input_node_names(node)
         global_state.update_aggregation_info(
-            query_graph=self.graph, node=node, input_node_names=input_node_names
-        )
-        to_decompose = global_state.should_decompose_query_graph(
-            node_name=node.name, input_node_names=input_node_names
+            query_graph=self.graph,
+            node=node,
+            input_node_names=input_node_names,
+            extract_primary_entity_ids_only=global_state.extract_primary_entity_ids_only,
         )
-        if to_decompose:
-            global_state.decompose_node_names.add(node.name)
-            global_state.update_ingest_graph_node_output_names(input_node_names=input_node_names)
-            global_state.node_name_to_aggregation_info[node.name].has_ingest_graph_node = True
+        if not global_state.extract_primary_entity_ids_only:
+            # skip decompose point extraction if the graph is only used to extract primary entity ids
+            # to speed up the process
+            to_decompose = global_state.should_decompose_query_graph(
+                node_name=node.name, input_node_names=input_node_names
+            )
+            if to_decompose:
+                global_state.decompose_node_names.add(node.name)
+                global_state.update_ingest_graph_node_output_names(
+                    input_node_names=input_node_names
+                )
+                global_state.node_name_to_aggregation_info[node.name].has_ingest_graph_node = True
         return global_state
 
     def extract(
         self,
         node: Node,
         relationships_info: Optional[List[EntityRelationshipInfo]] = None,
+        extract_primary_entity_ids_only: bool = False,
         **kwargs: Any,
     ) -> DecomposePointState:
-        # identify aggregation node names in the graph, aggregation node names are used to determine
-        # whether to start decomposing the graph
-        op_struct_state = OperationStructureExtractor(graph=self.graph).extract(node=node)
-        op_struct = op_struct_state.operation_structure_map[node.name]
-        aggregation_node_names = {agg.node_name for agg in op_struct.iterate_aggregations()}
+        aggregation_node_names = set()
+        operation_structure_map = {}
+        if not extract_primary_entity_ids_only:
+            # identify aggregation node names in the graph, aggregation node names are used to determine
+            # whether to start decomposing the graph
+            op_struct_state = OperationStructureExtractor(graph=self.graph).extract(node=node)
+            op_struct = op_struct_state.operation_structure_map[node.name]
+            aggregation_node_names = {agg.node_name for agg in op_struct.iterate_aggregations()}
+            operation_structure_map = op_struct_state.operation_structure_map
```

Comment:
Operation structure extraction is an expensive operation. Only use it when `extract_primary_entity_ids_only` is `False`.